ckeditor/ckeditor4
# Add preserving custom colors feature

## What is the purpose of this pull request?

New feature

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains


## What is the proposed changelog entry for this pull request?

```
* [#1795](https://github.com/ckeditor/ckeditor4/issues/1795): Colors picked from [Color Dialog](https://ckeditor.com/cke4/addon/colordialog) are now stored in [Color Button](https://ckeditor.com/cke4/addon/colorbutton) palette and can be easily reused.
* [#3783](https://github.com/ckeditor/ckeditor4/issues/3783): Colors used in the document are now displayed as a part of the [Color Button](https://ckeditor.com/cke4/addon/colorbutton) palette.
```

## What changes did you make?

* ## Features
Implementation is mainly based on conclusions listed [here](https://github.com/ckeditor/ckeditor4/issues/1795#issuecomment-559809627), but during development I made one significant change - in palette there are two separate sections for `Content Colors` (colors found in the editor content - checked on every `colorbutton` panel open) and `Custom Colors` (colors picked from color dialog). Without this division the palette was changing unintuitive when e.g. some parts of coloured text were deleted. 
Thanks to that it was also possible to implement both features independently, so theoretically this PR could be divided into two if it's easier to work with for reviewer.

* ## Unit tests
1. As we are in the middle of discussion about place of issue references in unit tests (using `bender-tags` or not), for now I just added references for the whole files as it was quicker way. If we determine that adding tags for each test is better, of course I'll change it.
2. Features are implemented for both `Text Color` and `Background Color` buttons and the only difference between them is that they should track colors independently, so only tests checking setup are doubled - the core features are tested only for `Text Color`. If this is insufficient, of course I'll change it.
3. Code for preserving color style is entirely in colorbutton plugin so I also put tests there, but in fact without colordialog it doesn't do a thing, so I wonder if it would be better to place them in the colordialog folder.

* ## Manual tests
1. I'm guess despite manual tests are split into 2 files they may still be to complex, if reviewer will share this point of view then of course I'll change it.
2. The same as point 3. for unit tests - maybe tests for custom colors should be in colordialog folder?

* ## PS
1. There will be probably two new phrases to translate - right now they are "Content Colors" and "Custom Color". Yet I didn't update lang files, as they can still change - let's create new entries just before merge.
2. In our test base I saw e.g. such code snippet:
```javascript
<h2 style="color:FF0000">Hello</h2>
```
and in this format the color will not be displayed in palette (which makes sense with because default ACF settings don't allow this construction).
As for now this feature only finds colors expressed in hex code attached to spans through `style` attribute, like here:
```javascript
<p><span style="color:334455">Hello</span></p>
```
But I just realised that the below snippet is also fine for ACF:
```javascript
<p><span style="color:red">instance</span></p>
```
so probably it should work too, and now it doesn't... So I'll fix that after the first review :)

## Which issues your PR resolves?

Closes #1795.
Closes #3783.
